### PR TITLE
Miracast fixes

### PR DESCRIPTION
**Pull request by thopiekar**
As discussed with @MarcKe on XDA, here are the patches I made to get a connection working to a WiDi adapter. The only problem is that the video stream hangs (framerate about 1~5 per minute), but this can be related to my cheap Miracast adapter.

What's still courious, is that I still get color format warnings in the logcat, while my build environment is up-to-date. I'm new at cyanogenmod/android development to I probably missed something.

```
W/ACodec ( 1936): do not know color format 0x7f000001 = 2130706433
W/ACodec ( 1936): do not know color format 0x7f000002 = 2130706434
W/ACodec ( 1936): do not know color format 0x7fc00002 = 2143289346
W/ACodec ( 1936): do not know color format 0x7f000003 = 2130706435
W/ACodec ( 1936): do not know color format 0x7f000789 = 2130708361
```

PS: I know these WFD_DEBUG sections are ugly, but gold when looking for a problem.

### CODE REVIEW

**Comment by MarcKe** [comment]
Ok, so basically you just make it ignore the hdcp protection.
We would need to define DEVICE_HAS_NO_HDCP_SUPPORT in our device tree, right?

**Comment by thopiekar** [comment]
Thats right, but the setup process of the connection will fail as the HW codec will throw an error.
So, in addtion to DEVICE_HAS_NO_HDCP_SUPPORT we will also need IGNORE_AVC_INTRA_MACROBLOCK_REFRESH_MODE_ERROR.

**Comment by thopiekar** [comment]
I tried fixing that by using Quirks in the "media-codecs.xml", but it didn't take any effect. I hope we'll find a better solution for that.

**Comment by MarcKe** [comment]
I cannot test this. Archi, do you have an adapater?

**Comment by MarcKe** [comment]
https://github.com/ArchiDroid/android_frameworks_av/pull/2

Just to add this :dancer:

**Comment by JustArchi** [comment]
I don't have one as well.

Interesting thing you found out tho.

**Comment by thopiekar** [comment]
Thanks, most of all I used this source https://github.com/kensuke/How-to-Miracast-on-AOSP/wiki/Source-patch .
There are also patches for the sink (in case you like to cast the screen to another phone/tablet).
The author of this (horrible) wiki and repository, also uploaded videos of his results.
https://www.youtube.com/watch?v=KNEmaIr3FEg
https://www.youtube.com/watch?v=IeXtIGuha80 (phone to phone)

So I still beleave that it does not work completely for me because of the upcoming color format warnings or because of my Miracast adapter itself.

**Comment by ZjemCiKolege** [comment]
Samsung smart tv with wifi direct suitable? P

**Comment by thopiekar** [comment]
It _should_ when it supports screen mirroring aka miracast.

**Comment by thopiekar** [comment]
The most recent commit should fix those misleading log messages :smiling_imp:

File: media/libstagefright/ACodec.cpp [hunk 1]
```diff
@@ -657,10 +657,10 @@ status_t ACodec::configureOutputBuffersFromNativeWindow(
     setNativeWindowColorFormat(eNativeColorFormat);
 
     err = native_window_set_buffers_geometry(
-    mNativeWindow.get(),
-    def.format.video.nFrameWidth,
-    def.format.video.nFrameHeight,
-    eNativeColorFormat);
+            mNativeWindow.get(),
```
Comment (JustArchi) [inline media/libstagefright/ACodec.cpp:660]: You should keep original format when editing source file

**Comment by thopiekar** [comment]
Sorry, I just found it better to see the differences between the patched and the unpatched section of this #if. Do I have to revert this?

Additionally, the possible fix of the unknown colour formats is useless and leads to a compilation error. Will fix that soon.

**Comment by JustArchi** [comment]
When you're done, just squash all those commits into one and make separate pull request so we can merge it in "nice" way

**Comment by thopiekar** [comment]
Ok, but in this case without formating the code?

**Comment by MarcKe** [comment]
https://github.com/ArchiDroid/android_hardware_samsung/commit/862a401e672c36084c31c5f8162569793e8d990e

Maybe this changes something regarding the intra refresh mode. I cannot test it.

**Comment by thopiekar** [comment]
Sounds good! Are these fixes now in Weekly 8?
Sorry, for being not active on that anymore, as I'm currently busy :pensive: 

PS: Also got random reboots the last days on my i9300. One day I even didn't care about my phone and found it before going to sleep uncharged while plugged to a charger (running at maximum) and the phone heated to >> 60°C (don't know exactly but not even smelling good :pensive: )
Have to do a hard reset and clean up now..

**Comment by ZjemCiKolege** [comment]
Still wifi direct no work :/ 
maybe pick this

https://github.com/NamelessRom/android_frameworks_av/commits/n-2.0 

https://github.com/NamelessRom/android_frameworks_av/commit/645d4f6824d66cfa066d32506d8c2ebecd890d88
https://github.com/NamelessRom/android_frameworks_av/commit/bb0050ed11de6148fc2d50af9ed3bdb800cba8c7
https://github.com/NamelessRom/android_frameworks_av/commit/8cf39326dda3350835e0dff344c8f0b884d5005b
https://github.com/NamelessRom/android_frameworks_av/commit/0251876fbefff85a45669f5e9b5e648db68dad98

I have no idea why it works on nameless and CyanogenMod does not work :+1:

**Comment by MarcKe** [comment]
Maybe check the repo first. Those commits are in and have nothing to do with it.

**Comment by MarcKe** [comment]
I could imagine some selinux denials, but I cannot test this.

**Comment by DimitriMayakovsky** [comment]
https://drive.google.com/file/d/0BzwpVlJ6FcBTc1VDUV9hbjlkLU0/view?usp=sharing

**Comment by thopiekar** [comment]
For me it works now like after my fixes when I skipped setting up the
"macro code". The connection is there, but video still at approx max 1
frame per minute and audio is still creepy.

Great that you found a friendly way to make it work like that, without
skipping everything that doesn't work ;)

However, when I find the time I'll push some cleanups I made when working
on that. Hope I won't forget it!
Am 16.03.2015 23:15 schrieb "MarcKe" notifications@github.com:

**Comment by DimitriMayakovsky** [comment]
Failed to set wfd properties

**Comment by ZjemCiKolege** [comment]
Oh some commits is hiden of nameless source maybe cast screen is in test :dancer:

**Comment by thopiekar** [comment]
@MarcKe The cheapest I found on ebay: http://www.ebay.de/itm/Ezcast-M2-OTA-Miracast-DLNA-WiFi-Display-Receiver-Dongle-Airplay-HDMI-1080P-/221713499284?pt=LH_DefaultDomain_146&hash=item339f278494
I heard some of them don't have Miracast implemented great, so I would suggest buying http://www.ebay.de/itm/heis-Vsmart-Ezcast-Wifi-Anzeige-Smart-TV-Stick-Media-Player-Dongle-DLNA-Airplay-/400838817344?pt=LH_DefaultDomain_77&hash=item5d53daf640 , if you like.
I own such a stick on my own - they have great support on Google+ and are able to update the firmware of-the-air.

**Comment by thopiekar** [comment]
Or: Go to the next multimedia store and do the hacks on their TV's. I'm sure they will be amused :smile:

**Comment by thopiekar** [comment]
Here is the log using Weekly8: http://pastebin.com/EDphEh58
Encoder setup begins with line 68.

**Comment by thopiekar** [comment]
@DimitriMayakovsky Build yourself a nightly with WFD_DEBUG, you might be able to see the reason why it fails
https://github.com/ArchiDroid/android_frameworks_av/commit/6b86fc508290d45284c61bb7b2ea4d6193cd9d85
https://github.com/ArchiDroid/android_frameworks_av/commit/30db14852416f2c8146df010b341c0609a02a859

**Comment by MarcKe** [comment]
Can you do me a favor?
Grab a kernel log. I want to see if selinux might be blocking it:
adb shell su -c dmesg | grep 'avc: '
You need to have root on adb activated.

**Comment by thopiekar** [comment]
I hate security: http://pastebin.com/xCU37GkK

**Comment by MarcKe** [comment]
You build on your own, right?
Try a build with this reverted (so like a revert of a revert :D): https://github.com/ArchiDroid/android_kernel_samsung_smdk4412/commit/e5e205a3057c737174e6b3ae8a231f14e3eb58c0

**Comment by thopiekar** [comment]
No, it is weekly 8 - totally clean install. Wiped everything because of my burning CPU (fixed after doing the wipe) and installed Weekly 8 + GApps.

**Comment by thopiekar** [comment]
But, yes, can build a nightly of course. Can you give me a `git` command to make the revert? And optional reenable it after the build?

**Comment by MarcKe** [comment]
Just use gitk or git revert e5e205a3057c737174e6b3ae8a231f14e3eb58c0 in the kernel directory

**Comment by MarcKe** [comment]
To reset you can just do repo sync again.

**Comment by thopiekar** [comment]
Disabled selinux via git revert and got the same here. Exactly the same behavior as I had with my fixes. It is just a suggestion, but the mirror casting uses the Android opaque color format to make the encoder choose a proper color format. When I read about our encoder on the net I read that ours (sometimes) can't handle it properly. This problem appeared with an unpatched camera app. Maybe we need such patches for miracast, too.

**Pull request closed without merging**